client/query_impl.go

```diff
@@ -0,0 +1,127 @@
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/iov-one/tendermint/rpc/client"
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/app"
+	"github.com/iov-one/weave/errors"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+func NewRPCABCIQuery(c client.Client) Query {
+	return &RPCABCIQuery{c}
+}
+
+// RPCABCIQuery exposes query interface via tendermint rpc
+type RPCABCIQuery struct {
+	conn client.Client
+}
+
+func (q *RPCABCIQuery) Get(key []byte) []byte {
+	query, err := q.conn.ABCIQuery("/", key)
+	if err != nil {
+		panic(err)
+	}
+
+	resp := query.Response
+	if resp.IsErr() {
+		panic(fmt.Sprintf("(%d): %s", resp.Code, resp.Log))
+	}
+
+	if len(resp.Key) == 0 {
+		return []byte{}
+	}
+
+	// assume there is data, parse the result sets
+	var keys, vals app.ResultSet
+	err = keys.Unmarshal(resp.Key)
+	if err != nil {
+		panic(err)
+	}
+	err = vals.Unmarshal(resp.Value)
+	if err != nil {
+		panic(err)
+	}
+
+	res, err := app.JoinResults(&keys, &vals)
+	if err != nil {
+		panic(err)
+	}
+
+	r, err := json.Marshal(res)
+	if err != nil {
+		panic(err)
+	}
+	return r
+}
+
+func (q *RPCABCIQuery) Iterator(start, end []byte) weave.Iterator {
+	panic("implement me")
+}
+
+func (q *RPCABCIQuery) ReverseIterator(start, end []byte) weave.Iterator {
+	panic("implement me")
+}
+
+func NewAppABCIQuery(app abci.Application) Query {
+	return &AppABCIQuery{app}
+}
+
+// AppABCIQuery exposes query interface via abci app query
+type AppABCIQuery struct {
```

Comment:
Q: And wondering, was there any discussion to write a client that interacts with ABCI apps directly? Which means effectively bypassing Tendermint? The only use case for that I can think of would be tests only. Given the amount of work with the parent ticket I would also depriotize this until the tendermint based client is completed.